### PR TITLE
Add type informations to torch.cuda

### LINKED ISSUE #47133: Enable torch.cuda typechecks during CI
`TestTypeHints.test_run_mypy` should pass with `[mypy-torch.cuda]` deleted from mypy.ini

cc @ezyang @malfet @rgommers @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Fixes #47133

### CODE REVIEW

File: torch/cuda/__init__.py [hunk 1]
```diff
@@ -467,8 +467,11 @@ class _CudaBase(object):
     is_sparse = False
 
     def type(self, *args, **kwargs):
-        with device(self.get_device()):
-            return super(_CudaBase, self).type(*args, **kwargs)
+        # We could use a Protocol here to tell mypy that self has `get_device` method
+        # but it is only available on Python >= 3.8 on typing or mypy_extensions on
```
Comment (rgommers) [inline torch/cuda/__init__.py:471]: I think you mean `typing_extensions`? That is a non-optional dependency already, so could be used. That said, I don't think it's necessary to spend time on this now. Just updating the comment may be good for now.

**Review by rgommers — APPROVED** [review]
Thanks @guilhermeleobas. The one CI failure that needed a closer look is `test_abs_cuda_complex128` for ROCm:

<details>

```
======================================================================
FAIL: test_abs_cuda_complex128 (__main__.TestForeachCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 834, in wrapper
    method(*args, **kwargs)
  File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 272, in instantiated_test
    result = test_fn(self, *args)
  File "test_foreach.py", line 259, in test_abs
    self.assertEqual(res, expected)
  File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1138, in assertEqual
    exact_dtype=exact_dtype, exact_device=exact_device)
  File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1112, in assertEqual
    self.assertTrue(result, msg=msg)
AssertionError: False is not true : Tensors failed to compare as equal! Attempted to compare equality of tensors with different dtypes. Got dtypes torch.float64 and torch.complex128.

======================================================================
FAIL: test_abs_cuda_complex64 (__main__.TestForeachCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 834, in wrapper
    method(*args, **kwargs)
  File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 272, in instantiated_test
    result = test_fn(self, *args)
  File "test_foreach.py", line 252, in test_abs
    self.assertEqual(res, expected)
  File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1138, in assertEqual
    exact_dtype=exact_dtype, exact_device=exact_device
```

</details>

It looks like a flake; that job has a different failure in many other PRs, but they're different each time.

Maybe just push a code comment update for my one other comment to check it's not the same failure twice?
LGTM other than that.

**Comment by rgommers** [comment]
Okay, ROCm failures are different ones now - these are all flakes.

File: torch/cuda/__init__.py [hunk 2]
```diff
@@ -467,8 +467,11 @@ class _CudaBase(object):
     is_sparse = False
 
     def type(self, *args, **kwargs):
-        with device(self.get_device()):
-            return super(_CudaBase, self).type(*args, **kwargs)
+        # We could use a Protocol here to tell mypy that self has `get_device` method
+        # but it is only available in the typing module on Python >= 3.8
+        # or on typing_extensions module on Python >= 3.6
```
Comment (ezyang) [inline torch/cuda/__init__.py:472]: btw we py3.6+ only now

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**